src/Common/OpenTelemetryTraceContext.cpp

```diff
@@ -14,76 +14,73 @@ namespace OpenTelemetry
 {
     if (!this->isTraceEnabled() || name.empty())
-        return;
+        return false;
 
-    this->attributes.push_back(Tuple{name, toString(value)});
+    return addAttribute(name, toString(value));
 }
 
-void Span::addAttributeIfNotZero(std::string_view name, UInt64 value)
+bool Span::addAttributeIfNotZero(std::string_view name, UInt64 value) noexcept
 {
-    if (value != 0)
-        addAttribute(name, value);
+    if (!this->isTraceEnabled() || name.empty() || value == 0)
+        return false;
+
+    return addAttribute(name, toString(value));
 }
 
-void Span::addAttribute(std::string_view name, std::string_view value)
+bool Span::addAttribute(std::string_view name, std::string_view value) noexcept
 {
     if (!this->isTraceEnabled() || name.empty())
-        return;
+        return false;
 
-    this->attributes.push_back(Tuple{name, value});
+    try
+    {
+        this->attributes.push_back(Tuple{name, value});
+    }
+    catch (...)
+    {
+        return false;
+    }
+    return true;
 }
 
-void Span::addAttributeIfNotEmpty(std::string_view name, std::string_view value)
+bool Span::addAttributeIfNotEmpty(std::string_view name, std::string_view value) noexcept
 {
-    if (!this->isTraceEnabled() || name.empty() || value.empty())
-        return;
-
-    this->attributes.push_back(Tuple{name, value});
+    return value.empty() ? false : addAttribute(name, value);
 }
 
-void Span::addAttribute(std::string_view name, std::function<String()> value_supplier)
+bool Span::addAttribute(std::string_view name, std::function<String()> value_supplier) noexcept
 {
     if (!this->isTraceEnabled() || !value_supplier)
-        return;
-
-    String value = value_supplier();
-    if (value.empty())
-        return;
-
-    this->attributes.push_back(Tuple{name, value});
-}
-
-void Span::addAttribute(const Exception & e) noexcept
-{
-    if (!this->isTraceEnabled())
-        return;
+        return false;
 
     try
     {
-        this->attributes.push_back(Tuple{"clickhouse.exception", getExceptionMessage(e, false)});
+        return addAttributeIfNotEmpty(name, value_supplier());
     }
     catch (...)
     {
-        /// Ignore exceptions
+        /// Ignore exception raised by value_supplier
+        return false;
     }
 }
 
-void Span::addAttribute(std::exception_ptr e) noexcept
+bool Span::addAttribute(const Exception & e) noexcept
+{
+    if (!this->isTraceEnabled())
+        return false;
+
+    return addAttribute("clickhouse.exception", getExceptionMessage(e, false));
+}
+
+bool Span::addAttribute(std::exception_ptr e) noexcept
 {
     if (!this->isTraceEnabled() || e == nullptr)
-        return;
+        return false;
 
-    try
-    {
-        this->attributes.push_back(Tuple{"clickhouse.exception", getExceptionMessage(e, false)});
-    }
-    catch (...)
-    {
-        /// Ignore exceptions
-    }
+    return addAttribute("clickhouse.exception", getExceptionMessage(e, false));
```

Comment:
`this->isTraceEnabled()` is checked twice: here + in `addAttribute` called in l. 83. What about introducing a method `bool Span::addAttribute_impl(std::string_view name, std::string_view value)` to which *all* `addAttribute()` methods map to? That way, you could implement all common checks only in a single place (_impl method).

Reply:
Good point. `addAttributeImpl`, a private method which has no parameter check, is added. All checks are still kept in the public methods this is because these checks can prevent extra functions calls such as `getExceptionMessage` if the checks return false.